viadot/sources/cloud_for_customers.py

```diff
@@ -0,0 +1,257 @@
+import re
+import requests
+import pandas as pd
+
+from copy import deepcopy
+from typing import Any, Dict, List, Optional
+from pydantic import BaseModel
+from urllib.parse import urljoin
+from viadot.exceptions import CredentialError
+
+from ..utils import handle_api_response
+from ..config import get_source_credentials
+from .base import Source
+
+
+class CloudForCustomersCredentials(BaseModel):
+    username: str  # eg. username@{tenant_name}.com
+    password: str
+
+
+class CloudForCustomers(Source):
+    """Cloud for Customers connector to fetch Odata source.
+
+    Args:
+        url (str, optional): The URL to the C4C API. E.g 'https://my396876543.jrt.ondemand.com/c4c/v1/'.
+        endpoint (str, optional): The API endpoint.
+        report_url (str, optional): The API URL in case of prepared report.
+        filter_params (Dict[str, Any], optional): Filtering parameters passed to the request. E.g {"$filter": "AccountID eq '1234'"}.
+        More info on: https://userapps.support.sap.com/sap/support/knowledge/en/2330688
+        credentials (CloudForCustomersCredentials, optional): Cloud for Customers credentials.
+        config_key (str, optional): The key in the viadot config holding relevant credentials.
+    """
+
+    DEFAULT_PARAMS = {"$format": "json"}
+
+    def __init__(
+        self,
+        url: str = None,
+        endpoint: str = None,
+        report_url: str = None,
+        filter_params: Dict[str, Any] = None,
+        credentials: CloudForCustomersCredentials = None,
+        config_key: Optional[str] = None,
+        *args,
+        **kwargs,
+    ):
+
+        ## Credentials logic
+        credentials = credentials or get_source_credentials(config_key)
+        if credentials is None:
+            raise CredentialError("Please specify the credentials.")
+        CloudForCustomersCredentials(**credentials)  # validate the credentials schema
+        super().__init__(*args, credentials=credentials, **kwargs)
+        ## End Credentials logic
+
+        self.url = url or self.credentials.get("url")
+        self.report_url = report_url
+
+        if self.url is None and report_url is None:
+            raise CredentialError("One of: ('url', 'report_url') is required.")
+
+        self.is_report = bool(report_url)
+        self.query_endpoint = endpoint
+
+        if filter_params:
+            filter_params_merged = self.DEFAULT_PARAMS.copy()
+            filter_params_merged.update(filter_params)
+
+            self.filter_params = filter_params_merged
+        else:
+            self.filter_params = self.DEFAULT_PARAMS
+
+        if self.url:
+            self.full_url = urljoin(self.url, self.query_endpoint)
+
+    @staticmethod
+    def create_metadata_url(url: str) -> str:
+        """Creates url to fetch metadata from.
```

Comment:
1) Let's be consistent, and I think "URL" is the correct grammar (like you use it on line 81 for example).
2) I didn't get the explanation TBH, I still don't know what this method does. I only get what `meta_url` is.

Reply:
1. Done
2. let us remember I didnt write this code so I am not fully 100% sure what this does.
But I think it is kind of clear that it is transforming the original url string, adding and removing stuff to the string, to change the composition of the url.
A simpler example containing the escence of this method would be:
def create_metadata_url(url: str = "https://mydomain.com/myendpoint"):
    meta_url = url + "metadata_endpoint/"
    return meta_url

Reply:
I could be wrong on this one, but I think it is practically what I have explained